### PR TITLE
Support bytes as 'to' address in transaction

### PR DESCRIPTION
**Pull request by stefanmendoza**
## What was wrong?

Issue #32 

Support for native `bytes` type as the `to` address when signing transactions.

## How was it fixed?

Added a check for instance of `bytes` with length of 20 before the `is_checksum_address` check when validating the fields in the transaction hash. Tested by adding another transaction dict in the accounts test with a `HexBytes` `to` attribute as well as adding a test to demonstrate the length-restriction on the `bytes`. Also commented some functional testing below.

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by stefanmendoza** [comment]
`master`:
```
(venv) [10:04:15 PM] @stefanmendoza ➜  eth-account git:(master) git last
commit 3ac5aed090aadd5f791f13ec23809b5b68bc4f08 (HEAD -> master, tag: v0.3.0, origin/master, origin/HEAD)
Author: Jason Carver <ut96caarrs@snkmail.com>
Date:   Tue Jul 24 12:12:31 2018 -0700

    Bump version: 0.2.3 → 0.3.0
(venv) [10:04:17 PM] @stefanmendoza ➜  eth-account git:(master) python
Python 3.6.5 (default, Jun 17 2018, 12:13:06) 
[GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from eth_account.account import Account
>>> from hexbytes import HexBytes
>>> acct1 = Account.create()
>>> acct2 = Account.create()
>>> baddr = HexBytes(acct2.address)
>>> transaction = { 'to': baddr, 'value': 1000000000, 'gas': 2000000, 'gasPrice': 234567897654321, 'nonce': 0, 'chainId': 1 }
>>> signed1 = Account.signTransaction(transaction, acct1.privateKey)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/stefanmendoza/Github/Forks/ethereum/eth-account/venv/lib/python3.6/site-packages/eth_utils/decorators.py", line 19, in _wrapper
    return self.method(objtype, *args, **kwargs)
  File "/Users/stefanmendoza/Github/Forks/ethereum/eth-account/eth_account/account.py", line 435, in signTransaction
    ) = sign_transaction_dict(account._key_obj, sanitized_transaction)
  File "/Users/stefanmendoza/Github/Forks/ethereum/eth-account/eth_account/internal/signing.py", line 23, in sign_transaction_dict
    unsigned_transaction = serializable_unsigned_transaction_from_dict(transaction_dict)
  File "/Users/stefanmendoza/Github/Forks/ethereum/eth-account/eth_account/internal/transactions.py", line 35, in serializable_unsigned_transaction_from_dict
    assert_valid_fields(transaction_dict)
  File "/Users/stefanmendoza/Github/Forks/ethereum/eth-account/eth_account/internal/transactions.py", line 140, in assert_valid_fields
    raise TypeError("Transaction had invalid fields: %r" % invalid)
TypeError: Transaction had invalid fields: {'to': HexBytes('0x77a6e9222f88997915220a561b1200e81773af36')}
>>> 
```

These changes:
```
(venv) [10:06:42 PM] @stefanmendoza ➜  eth-account git:(issue-32) git last
commit 3aaeac9b1ff74226e0958236d9eed48f1d41943a (HEAD -> issue-32, origin/issue-32)
Author: Stefan Mendoza <stefan@stefanmendoza.com>
Date:   Mon Oct 8 22:02:30 2018 -0500

    Support bytes as 'to' address in transaction
(venv) [10:06:45 PM] @stefanmendoza ➜  eth-account git:(issue-32) python
Python 3.6.5 (default, Jun 17 2018, 12:13:06) 
[GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from eth_account.account import Account
>>> from hexbytes import HexBytes
>>> acct1 = Account.create()
>>> acct2 = Account.create()
>>> baddr = HexBytes(acct2.address)
>>> transaction = { 'to': baddr, 'value': 1000000000, 'gas': 2000000, 'gasPrice': 234567897654321, 'nonce': 0, 'chainId': 1 }
>>> signed1 = Account.signTransaction(transaction, acct1.privateKey)
>>> signed1
AttrDict({'rawTransaction': HexBytes('0xf86a8086d55698372431831e848094a198151072b1f2e7888ab35c4a5d344f1b35e83c843b9aca008025a0ef42fe337f17c75131378a63a567c53ff511f08bde85722edec4398f5b5b90a9a05af63b756ea722f95aecb051436d7009763649749eeaa02c8a2bf57a269d523c'), 'hash': HexBytes('0x1458a662e5a487d73f362cc8e24ad7823cb4b2ee904169df7ed441571958ca9f'), 'r': 108221137149589847998295813862297888923307132406364953157283632778889175732393, 's': 41143211119430786298495731891179070660084854190859688004114680084137508688444, 'v': 37})
>>> type(acct2.address)
<class 'str'>
>>> transaction['to'] = acct2.address
>>> transaction['to']
'0xA198151072B1f2E7888ab35C4A5d344f1b35E83C'
>>> signed2 = Account.signTransaction(transaction, acct1.privateKey)
>>> signed1 == signed2
True
```

File: eth_account/internal/transactions.py [hunk 1]
```diff
@@ -66,12 +66,18 @@ def is_int_or_prefixed_hexstr(val):
...
         return True
+    elif is_valid_address_bytes(val):
+        return True
     elif is_checksum_address(val):
         return True
     else:
         return False
 
 
+def is_valid_address_bytes(val):
```
Comment (voith) [inline eth_account/internal/transactions.py]: This can be replaced by `is_binary_address` from `eth_utils`: 
`from eth_utils import is_binary_address `

Comment (stefanmendoza) [reply eth_account/internal/transactions.py]: Ah, good call! Will update that.

Comment (stefanmendoza) [reply eth_account/internal/transactions.py]: Fixed: https://github.com/ethereum/eth-account/pull/36/commits/139b9b77a1b9327daad724f192bbd96786600b32

**Comment by stefanmendoza** [comment]
@pipermerriam / @carver this is ready for review!

File: eth_account/account.py [hunk 2]
```diff
@@ -392,7 +392,6 @@ def signTransaction(self, transaction_dict, private_key):
         .. code-block:: python
 
             >>> transaction = {
-                    # Note that the address must be in checksum format:
```
Comment (carver) [inline eth_account/account.py]: Let's leave this message in, but add `or native bytes`. It's nice to hint to people who try to use a non-checksum address and get failures.

Comment (stefanmendoza) [reply eth_account/account.py]: Fixed!

https://github.com/ethereum/eth-account/pull/36/commits/abbae838ccb5297469e4df5fc266e575649939f6

**Comment by stefanmendoza** [comment]
@carver do we normally squash commits before merging?

**Comment by carver** [comment]
Often yes, but don't worry about this one. Also, we don't necessarily squash to a single commit, if there are natural separation of changes. I don't worry about squashing except for dirty fixup commits, or iteration on the same implementation where the intermediate versions are not worth keeping in history. Even then, I don't worry about an extra commit or two.

**Pull request merged**